File: kata3/python/kata.py

# def to_decimal_ip(inp):
#     while inp:
#         octet = inp[:8]
#         inp = inp[9:]
#         octet = [2 * int(x) for x in octet]


# def make_kata(ipaddress):
#     ip, mask = ipaddress.split('/')
#     binary_ip = ''.join(['%8s' % bin(int(x))[2:] for x in ip.split('.')]).replace(' ', '0')
#     assert len(binary_ip) == 32
#     mask = int(mask)
#     wildcard = 32 - mask
#     binary_subnet = binary_ip[:mask] + '0' * wildcard
#     binary_broadcast = binary_ip[:mask] + '1' * wildcard
#     binary_wildcard = '0' * mask + bin(wildcard)[2:]
#     subnet = []
#     broadcast = []

#     while binary_subnet:
#         octet = binary_subnet[:8]
#         binary_subnet = binary_subnet[8:]
#         octet = sum([int(mult) * 2 **(7 - pos) for pos, mult in enumerate(octet)])
#         subnet.append(octet)
#     subnet = '.'.join([str(x) for x in subnet])

#     while binary_broadcast:
#         octet = binary_broadcast[:8]
#         binary_broadcast = binary_broadcast[8:]
#         octet = sum([int(mult) * 2 **(7 - pos) for pos, mult in enumerate(octet)])
#         broadcast.append(octet)
#     broadcast = '.'.join([str(x) for x in broadcast])
#     return subnet, broadcast


# def make_kata(ipaddr):
#     subnet = broadcast = ''
#     ip, mask = ipaddr.split("/")
#     ip = ''.join([bin(int(x) + 256)[3:] for x in ip.split(".")])
#     mask = int(mask)
#     assert len(ip) == 32, "ip is %s; len(ip) is %s" % (ip, len(ip))
#     subnet = ip[:mask] + '0' * (32 - mask)
#     broadcast = ip[:mask] + '1' * (32 - mask)
#     assert len(subnet) == len(broadcast) == 32
#     def to_decimal_ip(binary_ip):
#         iplist = [binary_ip[x:x+8] for x in range(0, len(binary_ip), 8)]
#         for position, binary_octet in enumerate(iplist):
#             octet = 0
#             for pos, symbol in enumerate(binary_octet):
#                 octet += int(symbol) * 2 ** (7 - pos)
#             iplist[position] = str(octet)
#         return '.'.join(iplist)
#     return to_decimal_ip(subnet), to_decimal_ip(broadcast)


def make_kata(ipaddr):
    ip, mask = ipaddr.split("/")
    mask = int(mask)
    ip = ''.join([bin(int(octet) + 256)[3:] for octet in ip.split('.')])
    assert len(ip) == 32
    subnet = ip[:mask] + '0' * (32 - mask)
    broadcast = ip[:mask] + '1' * (32 - mask)

    def to_decimal_ip(ip):
        iplist = [ip[x:x + 8] for x in range(0, 32, 8)]
        for position, octet in enumerate(iplist):
            iplist[position] = str(sum([int(sym) * 2 ** (7 - pos) for pos, sym in enumerate(octet)]))
        return '.'.join(iplist)
    return to_decimal_ip(subnet), to_decimal_ip(broadcast)

File: kata3/kata.go
package kata3

import (
	"fmt"
	"math"
	"strconv"
	"strings"
)

type IpAddress struct {
	ip   [4]int
	mask int
}

func (self IpAddress) Stringer() string {
	return fmt.Sprintf("%v.%v.%v.%v/%v", self.ip[0], self.ip[1], self.ip[2], self.ip[3], self.mask)
}

func (self IpAddress) Binary() [32]bool {
	var result [32]bool
	ipString := fmt.Sprintf("%08b%08b%08b%08b", self.ip[0], self.ip[1], self.ip[2], self.ip[3])
	for i := 0; i < 32; i++ {
		if ipString[i] == '1' {
			result[i] = true
		} else {
			result[i] = false
		}
	}
	return result
}

func (self IpAddress) Subnet() IpAddress {
	binaryIp := self.Binary()
	for i := self.mask; i < 32; i++ {
		binaryIp[i] = false
	}
	return fromBinary(binaryIp, self.mask)
}

func (self IpAddress) Broadcast() IpAddress {
	binaryIp := self.Binary()
	for i := self.mask; i < 32; i++ {
		binaryIp[i] = true
	}
	return fromBinary(binaryIp, self.mask)
}

func FromString(in string) IpAddress {
	var ip [4]int
	ipSlice := strings.Split(in, "/")
	mask, _ := strconv.Atoi(ipSlice[1])
	for pos, val := range strings.Split(ipSlice[0], ".") {
		ip[pos], _ = strconv.Atoi(val)
	}
	return IpAddress{ip, mask}
}

func fromBinary(ip [32]bool, mask int) IpAddress {
	var integerIp [4]int
	mult := 7.0
	index := 0
	for pos, val := range ip {
		switch pos {
		case 8, 16, 24:
			mult = 7.0
			index += 1
		}
		if val {
			integerIp[index] += int(math.Pow(2.0, mult))
		}
		mult -= 1
	}
	return IpAddress{integerIp, mask}
}

func MakeKata(ipaddr string) (string, string) {
	ip := FromString(ipaddr)
	subnet := ip.Subnet()
	broadcast := ip.Broadcast()
	return subnet.Stringer(), broadcast.Stringer()
}
